DS-79 Assetstore to support different implementations, including S3

peterdietz opened the pull request:
https://jira.duraspace.org/browse/DS-79
A refactor of the Bitstream Storage Service to allow different implementations for file/bitstream storage. The primary motivation is leverage a cloud service, such as Amazon S3, which offers easy management, infinite scalability, extreme durability, and low pricing. Other storage implementations could be built off of this work. This follows the DSpace 2.0 goal of having a Pluggable Assetstore. Longsight is using this S3 bitstream storage feature for our cloud hosted DSpace instances on AWS.

Replaces the previous PR (leaving for posterity, and for those wanting a 5x-based branch) https://github.com/DSpace/DSpace/pull/960

~~This changes dspace.cfg for connecting assetstores.~~
Update: 11/26/2015, This PR now uses spring beans to declare and auto-configure each bitstore. All of the assetstore config has been yanked from dspace.cfg. Its actually less code to configure everything.

Configure bitstores in `config/spring/api/bitstore.xml`

```
<bean name="org.dspace.storage.bitstore.BitstreamStorageService" class="org.dspace.storage.bitstore.BitstreamStorageServiceImpl">
        <property name="incoming" value="1"/>
        <property name="stores">
            <map>
                <entry key="0" value-ref="localStore"/>
                <entry key="1" value-ref="s3Store"/>
            </map>
        </property>
    </bean>

    <bean name="localStore" class="org.dspace.storage.bitstore.DSBitStoreService" scope="singleton">
        <property name="baseDir" value="${dspace.dir}/assetstore"/>
    </bean>

    <bean name="s3Store" class="org.dspace.storage.bitstore.S3BitStoreService" scope="singleton">
        <property name="awsAccessKey" value="YOUR_ACCESS_KEY"/>
        <property name="awsSecretKey" value="ItsSecret"/>
        <property name="bucketName" value="dspace-asset-abcu"/>
        <property name="awsRegionName" value=""/>
        <property name="subfolder" value=""/>
    </bean>
```

== Using S3 Storage
Register at aws.amazon.com, and create an IAM user with policy permission to access/read/write to a specific S3 bucket, or S3FullAccess (loose permission). You will receive an access-key and secret-key, when you create a user. Then configure S3BitStoreService bean in config/spring/bitstore.xml

S3 will work just like a local assetstore, but requires internet connection between your DSpace server, and AWS. It is probably best to have your application server inside of AWS, so EC2, to minimize bandwidth costs and latency.
![Screenshot of AWS S3 admin console, view of assets](https://cloud.githubusercontent.com/assets/58014/8016204/6ed36fbc-0bae-11e5-99f9-bc90100c4a80.png)

This refactor of BitstreamStorageManager is based on Richard Rodger's work of Pluggable Storage
https://wiki.duraspace.org/display/DSPACE/DSpace+2.0+Pluggable+Storage

peterdietz committed:
Change BitstreamStorageService to use BitStore interface, adding S3 storage

peterdietz committed:
DS-2863 Stop checksum checker from crashing

peterdietz committed:
Fix checksum findBitstreams with not recent checksum

peterdietz committed:
Update dspace.cfg with updated assetstore config

peterdietz commented:
This also includes a tool to migrate assets between bitstores. So, you can move all bitstreams from assetstore[0] to assetstore[1], independent of implementation.

```
/dspace/bin/dspace bitstore-migrate
usage: BitstoreMigrate
 -a,--source <arg>        Source assetstore store_number (to lose content). This is a number such as 0 or 1
 -b,--destination <arg>   Destination assetstore store_number (to gain content). This is a number such as 0 or 1.
 -d,--delete              Delete file from losing assetstore. (Default: Keep bitstream in old assetstore)
 -h,--help                Help
 -p,--print               Print out current assetstore information
 -s,--size <arg>          Batch commit size. (Default: 1, commit after each file transfer)

/dspace/bin/dspace bitstore-migrate -p
store[0] == DSBitStore, which has 2 # of bitstreams.
store[1] == S3BitStore, which has 2 # of bitstreams.
Incoming assetstore is store[1]

/dspace/bin/dspace bitstore-migrate -a 0 -b 1

/dspace/bin/dspace bitstore-migrate -p
store[0] == DSBitStore, which has 0 # of bitstreams.
store[1] == S3BitStore, which has 4 # of bitstreams.
Incoming assetstore is store[1]
```

helix84 commented:
Hi @peterdietz, just make it explicitly clear, this removes SRB, correct?

peterdietz commented:
@helix84 Yes, this removes SRB support. I made an announcement on [dspace-tech](https://groups.google.com/forum/#!searchin/dspace-tech/srb%7Csort:date/dspace-tech/sedtn52A7xg/MK-cstgEBwAJ) several months ago seeking input from SRB users, but none stood up. This refactor brings a clean interface in BitStore, that adding an SRB implementation should be simple. We could try to just copy over the SRB related code into an implementation, but since nobody has access to SRB, then we couldn't verify that it is still correct.

Also, the SRB implementation has a really bad checksum mechanism. It just computes a checksum from the filename, as opposed to computing it over the full binary file.

peterdietz committed:
Fix DSBitstore register bitstream

KevinVdV commented on dspace-api/src/main/java/org/dspace/checker/MostRecentChecksum.java:
```diff
@@ -61,6 +64,35 @@
     @JoinColumn(name= "result")
     private ChecksumResult checksumResult;
 
+    public MostRecentChecksum(Bitstream bitstream) {
```
I would move this to the MostRecentChecksumService, we want to keep our Data objects as clean as possible.

KevinVdV commented:
Checked the code & just have one questions/recommendations. Would it not be better to use Spring to configure the assetstore implementations ?
It might be a cleaner way & will result in less code were you manually have to retrieve the properties & create the DSBitStore objects.

I'm also wondering if it would be better to rename DSBitStore to DSBitStoreService (because it is technically a service that takes care of how you store you data)

Haven't tested anything yet, this was just a code review.

tdonohue commented:
I agree with @KevinVdV that we might want to change some class names here just to make them consistent with the new style API. It's a tiny thing though.

A few other minor things:
* We'll need to remove documentation on SRB. I do agree with it's removal here though
* Needs its own Wiki Docs (obviously), with some explicit warnings about possible costs/latency if you are not using AWS for your DSpace Server. (Those are mentioned above, but we should be sure to mention in our documentation)
* We should open up a separate ticket documenting the removal of SRB, and close it as fixed by this PR (once merged).

mwoodiupui commented on dspace/config/dspace.cfg:
```diff
@@ -107,77 +107,37 @@ mail.allowed.referrers = ${dspace.hostname}
+# Other implementations take path to configuration file relevant to that method
+## assetstore.<#>=<short-name>:<param>
+assetstore.0=ds:${dspace.dir}/assetstore
+assetstore.1=s3:${dspace.dir}/config/modules/bitstore-s3.cfg
```
Should be commented -- most sites won't have S3 yet.

tdonohue commented on dspace/config/dspace.cfg:
```diff
@@ -107,77 +107,37 @@ mail.allowed.referrers = ${dspace.hostname}
+# Configure the bitstore implementations
+## bitstore.<short-name>.class=<implementation-class-canonical-name>
+bitstore.ds.class=org.dspace.storage.bitstore.DSBitStore
+bitstore.s3.class=org.dspace.storage.bitstore.S3BitStore
```
As noted by @KevinVdV, it's arguable these should just be done via Spring. We should avoid specifying class names in our config files as much as possible, and move these into out-of-the-box Spring configs

mwoodiupui replied:
Then we should move the *entire* assetstore configuration into Spring.  Each BitStore instance injected will need to have a number as part of its configuration -- these numbers must be stable because they are used in Bitstream.  The assetstore to receive new bitstreams should also be injected alongside the BitStore instances so that all of the information is collected in one spot.

mdiggory replied:
We do already have similar solutions internally. An ideal configuration would look something like the following.

```
    <bean name="org.dspace.storage.bitstore.BitstreamStorageService" class="org.dspace.storage.bitstore.BitstreamStorageServiceImpl">
        <property name="defaultIncomingBitStore" value="0"/>
        <property name="bitStores">
            <map>
                <entry key="0" value-ref="localStore"/>
            </map>
        </property>
    </bean>

    <bean name="localStore" class="org.dspace.storage.bitstore.storeimpl.DSBitStore" scope="singleton">
        <property name="baseDir" value="${dspace.dir}/assetstore"/>
    </bean>
```

mwoodiupui commented:
Lightly tested with normal submission and (especially) registration.  +1

peterdietz committed:
Disable s3storage from defaults

peterdietz committed:
Log all exceptions from DSBitstore

peterdietz committed:
Rename BitStore to BitStoreService, and DSBitStore, S3BitStore

peterdietz committed:
Move assetstore configuration to bitstore service spring beans

Remove configuration of assetstore from dspace.cfg, or bitstore-s3.cfg,
and instead magically configure this through config/spring/api/bitstore.xml

peterdietz commented:
Thanks all for the feedback/testing. I have renamed BitStore (and others) to be named service, like BitStoreService.

I have wired bitstore as a spring bean, it actually cleaned up code handling configuration and init much more than I expected. Look at `config/spring/api/bitstore.xml`.

I haven't looked too deeply into Guava libraries for resume/stream, but I feel if you change the signature, you'll have to propogate that up to how each UI/webservice exposes content.

Next, need to finish fixing Checksum/MostRecentChecksum.

peterdietz commented:
This PR has attempted to touch Checksum / MostRecentChecksum. It doesn't break checksums, but it doesn't fix checksums (its broken in master). That ticket is here: https://jira.duraspace.org/browse/DS-2863 I will either need to finish fixing checksums (in this PR), or yank out the commits that have tried to fix that from this PR, and make a separate PR for checksum issues.

rradillen commented:
Did you implement any mitigations for preventing filter-media from needing to download all files. 
This is also a concern for the checksum checker. (I think aws support this on the server side using  If-None-Match http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectGET.html http://docs.aws.amazon.com/AmazonS3/latest/API/RESTCommonResponseHeaders.html). 

It seems a bit wasteful to retrieve the bitstream from S3 to send it on to the user while the pipe between the user and S3 is probably much wider than the pipe between the user and any of our servers. Giving the user a signed link would be preferable I would think. Of the top of my head I see two potential issues. 1 How to expose this to the upper layers without breaking abstraction. 2. how to enforce bitstream rights resource policy. Does a signed link with a short lifetime suffice?

peterdietz commented:
Hi Roeland, yes, sending the data directly from S3 to the end user is more efficient. And, short-lived signed url would probably be a valid route to go. But, I've had the same thought as you, being able to expose that information will likely cause a big change to the abstraction layers. Also, I wonder if GoogleBot would be happier to download PDF's at a stable url, like dspace . domain . edu /bitstream/1234/thesis.pdf, as opposed to s3 . aws . com / bucket-name / dynamic-nonce-secret-1234-!@#$%^&*() / thesis.pdf

I haven't made any specific optimizations to media-filter, checksum, or index-discovery to prevent them from having to pull too much IO. But yes, I would assume that the current DSpace algo for this is naive, and probably needlessly does more IO than necessary.

The final piece of changes that has been suggested to this PR before it can be merged is to yank out the checksum-checker fixes, and have that be a separate issue.

peterdietz committed:
Move MostRecentChecksum logic to service

peterdietz committed:
Fix checksum history command by creating object first, then create

peterdietz committed:
Add bitstreamStorage.computeChecksum method, to replace checksum retrieve

peterdietz committed:
Changed hibernate field for checker-emailer

peterdietz commented:
Hi @rradillen thanks for your message. I looked into this code more. I was able to fix the checksum code. So instead of yanking out attempts to fix checksum, I just kept going, and have checksum checking more correct/fixed now. 

Also, for performance sake, I did look for where checksum code was running RETRIEVE on every file, and then doing its own digest computing code. I've replaced that will a new method BitstreamStorage.computeChecksum, which uses the bitstore's about(bitstream) method, which returns "technical metadata", which is usually size_bytes, checksum, checksum_algorithm. AWS S3 in about method can do a remote s3-side md5 checksum check (this method is called getETag ). So, instead of moving 100's of Gigabytes a day, just to run checksum checker !!!!! This will just be minor metadata overhead, and no binary retrieves, atleast for s3. The local DSBitstore has to be computed locally, so that will run as normal.

So. I would appreciate a bit more testing. Checksum works better, probably in the case of using s3, it will be very fast, and yet I assume still accurate.

peterdietz committed:
Remove public constructor for MostRecentChecksum after rebase

peterdietz commented:
I have just rebased this branch to fix merge conflicts. (MostRecentChecksum and ChecksumHistory classes were touched in master)

hardyoyo commented:
Tested in Vagrant-DSpace, I can confirm that the bitstore-migrate script works as described in the draft documentation. After migration, bitstreams still work as they should. Submitting a new file works as well. +1

peterdietz commented:
Documentation for updated storage layer is available in: https://wiki.duraspace.org/display/DSDOC6x/Storage+Layer
